### DISCUSSION TITLE
Pass email address as additional information to QR code

### DISCUSSION THREAD

**Discussion by mkue** [schoero/SwissQRBill#437] [General]
Hello👋

First of all, thank you for your efforts in creating and maintaining this library!

I'm trying to pass an email address as additional information to the QR code payment so that I can match the payment to a user when I retrieve the XML with the payment details from the bank.

This is how I created a QR code to test the process:

```ts
new SwissQRCode({
	amount: 1,
	currency: 'CHF',
	creditor: {
		account: 'CH6730000001151126386',
		name: 'Social Income',
		address: 'Zweierstrasse',
		buildingNumber: 103,
		city: 'Zürich',
		zip: 8003,
		country: 'CH',
	},
	av1: 'av1: test@test.org',
	av2: 'av2: test@test.org',
	message: 'Contribution by test@test.org',
	reference: '000000017013512241470000010',
});
```

I used the generated QR code to then make a payment. The `RmtInf` section in the resulting payment XML document looked like this:

```xml
<RmtInf>
    <Strd>
        <CdtrRefInf>
            <Tp>
                <CdOrPrtry>
                    <Prtry>QRR</Prtry>
                </CdOrPrtry>
            </Tp>
            <Ref>000000017013512241470000010</Ref>
        </CdtrRefInf>
        <AddtlRmtInf>?REJECT?0</AddtlRmtInf>
        <AddtlRmtInf>?ERROR?000</AddtlRmtInf>
        <AddtlRmtInf>Contribution by test test.org</AddtlRmtInf>
    </Strd>
</RmtInf>
```

As you can see, there are two error messages `?REJECT?0` and `?ERROR?000`, plus the `@` character in the message also disappeared (I assume that the bank does not allow certain characters in the message…).

Does anybody know how to properly and correctly encode an email address in a QR bill? I would assume that this should be possible. The only thing I can come up with is to encode the email address in [base58](https://github.com/bitcoinjs/bs58check) (all characters are supported in QR bill) and add it as a message. But this just seems overly complicated, and the person paying the bill would see uninterpretable information in the message field of the payment. 

Thanks for any tips and help!

**Comment by schoero** [comment]
The `av1` and `av2` properties are `alternative schemes` ([4.3.4 in the specifications](https://www.six-group.com/dam/download/banking-services/standardization/qr-bill/ig-qr-bill-v2.2-en.pdf)). Those are fields which have a predefined structure. Currently the [only valid alternative schemes are TWINT and eBill](https://www.six-group.com/en/products-services/banking-services/payment-standardization/standards/qr-bill.html#scrollTo=alternative-schemes).

The purpose of the property `additionalInformation` is to define payment instructions for the recipient, ([Annex E in the specifications, referred to as `Billing Information`](https://www.six-group.com/dam/download/banking-services/standardization/qr-bill/ig-qr-bill-v2.2-en.pdf)). This data is not forwarded with the payment but it is printed on the payment part.

I think the field you want to use is the [`message`](https://github.com/schoero/swissqrbill/blob/main/docs/pdf/types.md#interface-data) property. As far as I am aware this field is unstructured and forwarded with the payment. 

You can also check the [QR-Bill validator](https://swiss-qr-invoice.org/validator/?lang=de) and drop in your generated PDF to view most errors.

**Reply by mkue** [reply]
Thanks a lot for your quick response and the tip with the QR-Bill validator.

The message did partially work in my example, but it only supports a limited set of characters , e.g. `@` was replaced with a space. Most likely other characters that are allowed in email addresses, like ä/ü/ö, will be replaced with a string too, which is a problem. It seems that using an encoding like base58 to encode the email is the safest approach then. Seems kind of weird to me, but that's how I'll do it for now...

Anyway, thanks again for your help!

**Reply by schoero** [reply]
The intended way to match a payment to a user is usually the reference number. You could keep a record of the reference numbers and the corresponding email addresses.

You are correct about the limited character set. I don't think that encoding it is a good idea because it will also get printed on the payment part as you have already pointed out. If you only use the QR Code as you have posted in your code example, you might get away with this.

I would recommend you to use the reference number if this is possible for you as the `message` field is not really intended for this use case.

Maybe I should also add a warning if other characters are used within the QR bill.
